### ISSUE TITLE
Unsoundness around translation of Sequences in Java?

### ISSUE THREAD

**Issue by DavePearce** [dafny-lang/dafny#2859]
I'm assuming this is already a known issue, though I did not find anything about it yet.  The problem is that verified Dafny code can fail at runtime with an `ArithmeticException`.  This is arising because of a conversion from a `BigInteger` to an `int`.  Fundamentally, however, this is because arrays in Java cannot exceed `2^31` bytes in size.

To demonstrate the problem, we have this method in a module `Bytes` which verifies without trouble:

```Dafny
function method Padding(n:nat) : seq<u8>
ensures |Padding(n)| == n
{
     seq(n, i => 0)
}
```

Having translated this into Java we can make the following call to cause an `ArithmeticException`:

```Java
Bytes_Compile.__default.Padding(BigInteger.valueOf(Long.MAX_VALUE));
```

The trace is:

```
java.lang.ArithmeticException: BigInteger out of int range
	at java.base/java.math.BigInteger.intValueExact(BigInteger.java:4784)
	at dafny.DafnySequence.Create(DafnySequence.java:123)
	at Bytes_Compile.__default.Padding(__default.java:68)
	...	
```

**Comment by DavePearce** [comment]
Thinking about it, I thought I might just clarify this a little further.  My example above is perhaps not the best because we would expect it to fail regardless (i.e. with an out of memory error).  However, we can refine it a little:

```
Bytes_Compile.__default.Padding(BigInteger.valueOf(Integer.MAX_VALUE+1L));
```

We are attempting to allocate `0x80000000` bytes (`2GB`) which, on my machine with with `32GB` is perfectly reasonable.  As a simple demonstration, the following does indeed correctly allocate this amount of memory:

```C
#include <stdio.h>                                                                                                                                                                                            
#include <stdlib.h>                                                                                                                                                                                           
                                                                                                                                                                                                              
void main(int argv, char**args) {                                                                                                                                                                             
  void *ptr = malloc(0x80000000);                                                                                                                                                                             
  if(ptr != NULL) {                                                                                                                                                                                           
    printf("Allocation Succeeded\n");                                                                                                                                                                         
  }                                                                                                                                                                                                           
} 
```

Whilst I do understand that memory allocation is a serious challenge here, it seems to me that we should expect the code to succeed when it reasonably can.  Otherwise, we are saying Dafny can fail without warning at any allocation, and that seems problematic to me.

**Comment by cpitclaudel** [comment]
Hi Dave, sorry to see that we haven't answer about this.  I suspect you know a lot of the following, but I'm writing this up for future readers too:

System limits like this are a persistent annoyance.  In general, Dafny's verification is done against an idealized machine model, and we don't try to provide stronger guarantees that the platform that we compile to.  This means that, in particular:

- We don't eliminate recursion, so on platforms with small stacks you may see stack overflows, and you may see e.g. C# crashing while Go returns successfully.
- We don't enforce a specific garbage collection policy, so code that runs out of heap space in C# may succeed in Java, or vice versa.
- We don't try to flag excessively large allocations.  If the language supports allocating a certain size we will attempt to, and depending on the platform you may see different results.

Note that you don't need a giant allocation to run into the issue: doing very many small allocations (and keeping references to them) will eventually crash your program (again, at different speeds depending on the platform, in particular depending on how much your system accepts to page and to overcommit memory)

Focusing on array allocations: what we do when you attempt to allocate an array is convert size to the largest type that can hold an array size in the target language, and try to allocate that.

Answering the specific points you made:

> Having translated this into Java we can make the following call to cause an ArithmeticException:

This is expected: the number doesn't fit in an int, and hence Java won't let you allocate an array that large.  The error message could be improved (to mention memory instead of arithmetic, or maybe just "platform limitations" instead of arithmetic).

> However, we can refine it a little:
> ```
> Bytes_Compile.__default.Padding(BigInteger.valueOf(Integer.MAX_VALUE+1L));
> ```
> We are attempting to allocate 0x80000000 bytes (2GB) which, on my machine with with 32GB is perfectly reasonable. 

Yes in general, but not in Java: https://stackoverflow.com/questions/31382531/why-i-cant-create-an-array-with-large-size

>  As a simple demonstration, the following does indeed correctly allocate this amount of memory:

Indeed, and the following Dafny program runs fine when compiled to C++, too:

```
ghost const TWO_TO_THE_64: int := 0x1_00000000_00000000

newtype u8 = i: int | 0 <= i < 256
newtype u64 = i: int | 0 <= i < TWO_TO_THE_64

method Padd(n: u64) returns (a: array<u8>)
  ensures a.Length == n as int
{
  a := new u8[n];
}

method Main() {
  var _ := Padd(0x1_00000000 as u64);
}
```

… but Java has different limitations, and that's what you're hitting in this case.

**Comment by DavePearce** [comment]
Hey Clément,

Thanks for the response.  Mostly I understand the issue, and I don't generally have any obvious work arounds.  This does seem like a critical issue for Dafny, however, if it wants to be used for developing real applications.  I guess that the original intention for Dafny was that it be used as teaching language (in some sense), and this has impacted its language design.

Anyway, there is only one point address in what you've said regarding allocating arrays in Java:

> Yes in general, but not in Java: 

I appreciate that this is true for _arrays_ in Java.  However, in fact, we are allocating a `DafnySequence` here.  Therefore, it is not limited to the allocation size of arrays.  For example, we could implement it using an array of arrays which would dramatically increase the maximum size that could (in theory) be allocated.  This would come at a cost of performance.

The real problem here I think is that this is not actually a resource allocation issue.  Rather, its a legacy issue caused by Java limiting the size of arrays to 32bits.  Anyways, there we go.  I'd be interested if there has been any talk on ways the language could be updated to address this.

**Comment by keyboardDrummer** [comment]
> I guess that the original intention for Dafny was that it be used as teaching language (in some sense), and this has impacted its language design.

What makes you think there's a flaw in the existing design, as opposed to certain features not being there yet?

> I'd be interested if there has been any talk on ways the language could be updated to address this.

Clement mentioned three issues: lack of stack space, lack of heap space, and allocations that are too large. I would argue that there's another similar issue here, which is that Dafny programs might not finish within the time that you need them to. All of these are situations in which Dafny doesn't guard against you going over the resource limits of the system.

I think the potential solution is a static analysis that determines how much your Dafny program uses these various resources, and that enables Dafny to pick and configures a runtime system that can handle that usage. You might have to let your program return early when it receives inputs that use too much resources, so the required resources are bounded.

I feel such an analysis is quite a while away from being added to Dafny, but it would be amazing to have.

**Comment by DavePearce** [comment]
> What makes you think there's a flaw in the existing design, as opposed to certain features not being there yet?

Well, I don't necessarily think there is a flaw.  Only that these issues didn't need to be considered in the original design.  That's actually the same for Whiley, in fact.  Having said that, resolving issues around memory presents some real challenges.

> Clement mentioned three issues: lack of stack space, lack of heap space, and allocations that are too large. All of these are situations in which Dafny doesn't guard against you going over the resource limits of the system.

Right, but my point is that this particular example is slightly different.  Its an _artificial_ resource limit imposed by Dafny which could (in this case) be resolved by modifying Dafny's Java runtime.  That is, rewriting `DafnySequence<>` to use an array of arrays rather than a single array.  Its a simple fix that would remove this artificial resource limit.  

Anyway, I'm not proposing that you make this change.  It was just an observation.

**Comment by cpitclaudel** [comment]
>  That is, rewriting DafnySequence<> to use an array of arrays rather than a single array. Its a simple fix that would remove this artificial resource limit.

Wouldn't it just increase the limit (rather than remove it entirely)?
Changing to a linked list of arrays would remove the limit (but we would run into OOM exceptions instead)

**Comment by DavePearce** [comment]
> Changing to a linked list of arrays would remove the limit (but we would run into OOM exceptions instead)

That's right.  But it means, when there is enough memory, the program would run.  Whereas at the moment, that is not the case.  

This is actually affecting our project right now.  We cannot execute certain tests because they trigger this exception.  These tests are not written by us, and are out of our control.  They deliberately push memory allocation to the 4G mark.  To resolve this, I have two choices: use a `seq<seq<byte>>` instead of just `seq<byte>` on the Dafny side; or, perhaps, implement my own version of `DafnySequence`.  The former is particularly unappealing because it means I'm hacking my Dafny code to work around a Java problem.  The latter is perhaps more reasonable, and I just write my own version of the runtime library.  

Anyways, I appreciate that this is not a priority for you and that this is really my problem to sort out.  However, I think it was at least worth some discussion around it, which we've had.
